### PR TITLE
UTs: run on dotnet 8.0.5

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
dotnet 8.0.5 introduced breaking changes when generating code for Razor files.

This PR updates our pipeline to:
- run UTs on 8.0.5
- describe the new behavior

Fixes: https://github.com/SonarSource/sonar-dotnet/issues/9291

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/MetricsAnalyzerTest.cs [hunk 1]
```diff
@@ -69,14 +69,14 @@ public class MetricsAnalyzerTest
                     var metrics = messages.Single(x => x.FilePath.EndsWith(RazorFileName));
 
                     metrics.ClassCount.Should().Be(1);
-                    metrics.CodeLine.Should().BeEquivalentTo(new[] { 3, 5, 8, 10, 13, 15, 16, 17, 19, 22, 23, 24, 26, 28, 29, 32, 33, 34, 36, 37, 39, 40, 43 });
+                    metrics.CodeLine.Should().BeEquivalentTo(new[] { 2, 1, 3, 5, 8, 10, 13, 15, 16, 17, 19, 22, 23, 24, 26, 28, 29, 32, 33, 34, 36, 37, 39, 40, 43 });
                     metrics.CognitiveComplexity.Should().Be(3);
                     metrics.Complexity.Should().Be(4);
-                    metrics.ExecutableLines.Should().BeEquivalentTo(new[] { 3, 5, 13, 15, 17, 24, 28, 29, 32, 36, 39, 43 });
+                    metrics.ExecutableLines.Should().BeEquivalentTo(new[] { 13, 15, 28, 29, 32, 36 });
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/MetricsAnalyzerTest.cs:75]: See also the comment from `CSharpExecutableLinesMetricTest.cs`. Arguments are no longer considered executable lines.

File: analyzers/tests/SonarAnalyzer.TestFramework.Test/Common/SourceGeneratorProviderTest.cs [hunk 2]
```diff
@@ -26,20 +26,20 @@ namespace SonarAnalyzer.TestFramework.Test.Common;
 public class SourceGeneratorProviderTest
 {
     private static AnalyzerFileReference RazorSourceGenerator =>
-        SourceGeneratorProvider.SourceGenerators.Single(x => x.FullPath.EndsWith("Microsoft.CodeAnalysis.Razor.Compiler.SourceGenerators.dll"));
+        SourceGeneratorProvider.SourceGenerators.Single(x => x.FullPath.EndsWith("Microsoft.CodeAnalysis.Razor.Compiler.dll"));
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework.Test/Common/SourceGeneratorProviderTest.cs:29]: The old library is no longer present. The source generators are now part of the `Microsoft.CodeAnalysis.Razor.Compiler.dll`.

File: analyzers/tests/SonarAnalyzer.TestFramework.Test/TestCases/Dummy.SecondaryLocation.razor [hunk 3]
```diff
@@ -1,7 +1,7 @@
 
-<p>The solution to all problems is: @(RaiseHere(21))</p> @* wrong location *@
-@*                      ^^^^^^^^^ *@
-@*                                ^^ Secondary@-1 *@
+<p>The solution to all problems is: @(RaiseHere(21))</p>
+@*                                    ^^^^^^^^^ *@
+@*                                              ^^ Secondary@-1 *@
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework.Test/TestCases/Dummy.SecondaryLocation.razor:4]: The locations are now correct 🎉

File: analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/MetricsAnalyzerTest.cs [hunk 4]
```diff
@@ -69,14 +69,14 @@ public class MetricsAnalyzerTest
                     var metrics = messages.Single(x => x.FilePath.EndsWith(RazorFileName));
 
                     metrics.ClassCount.Should().Be(1);
-                    metrics.CodeLine.Should().BeEquivalentTo(new[] { 3, 5, 8, 10, 13, 15, 16, 17, 19, 22, 23, 24, 26, 28, 29, 32, 33, 34, 36, 37, 39, 40, 43 });
+                    metrics.CodeLine.Should().BeEquivalentTo(new[] { 2, 1, 3, 5, 8, 10, 13, 15, 16, 17, 19, 22, 23, 24, 26, 28, 29, 32, 33, 34, 36, 37, 39, 40, 43 });
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/MetricsAnalyzerTest.cs:72]:
```suggestion
                    metrics.CodeLine.Should().BeEquivalentTo(new[] { 1, 2, 3, 5, 8, 10, 13, 15, 16, 17, 19, 22, 23, 24, 26, 28, 29, 32, 33, 34, 36, 37, 39, 40, 43 });
```

File: analyzers/tests/SonarAnalyzer.Test/TestCases/Utilities/MetricsAnalyzer/Razor.razor [hunk 5]
```diff
@@ -1,8 +1,8 @@
-@page "/razor"                                     // FN: line of code, comment (location mapping issue)
-@namespace RazorNamespace                          // FN: line of code, comment (location mapping issue)
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/Utilities/MetricsAnalyzer/Razor.razor:2]: Why those comments were removed?

File: analyzers/tests/SonarAnalyzer.Test/TestCases/Utilities/MetricsAnalyzer/Razor.razor [hunk 6]
```diff
@@ -37,10 +37,10 @@
 {                                                  // code line
     // comment
     @:Name: @i                                     // code line, FN: comment 
-}                                                  <!-- code line, FN: comment -->
+}                                                  <!-- FN: comment,code line -->
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/Utilities/MetricsAnalyzer/Razor.razor:40]: Educational: what is the `code line` comment means?

File: analyzers/tests/SonarAnalyzer.TestFramework.Test/TestCases/DummyExpressions.razor [hunk 7]
```diff
@@ -16,16 +16,16 @@
...
-<p>Lambda expression: @((Func<int>)(() => RaiseHere()))()</p> @* Noncompliant, wrong location *@
-@*                                         ^^^^^^^^^ *@
+<p>Lambda expression: @((Func<int>)(() => RaiseHere()))()</p> @* Noncompliant *@
+@*                                        ^^^^^^^^^ *@
 
 <p>Nested multi-statement block: @{ var result2 = RaiseHere(); var result3 = RaiseHere(); }</p>
 @*                                                ^^^^^^^^^ *@
 @*                                                                           ^^^^^^^^^@-1 *@
 
 <p>Nested control structures: @if(RaiseHere() == 42) { <text>@(RaiseHere() + 1)</text> }</p> @* second issue has wrong location *@
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.TestFramework.Test/TestCases/DummyExpressions.razor:26]:
```suggestion
<p>Nested control structures: @if(RaiseHere() == 42) { <text>@(RaiseHere() + 1)</text> }</p>
```

File: azure-pipelines.yml [hunk 8]
```diff
@@ -248,6 +248,12 @@ stages:
             FrameworkMoniker: 'net8.0'
 
       steps:
+      - task: UseDotNet@2
+        displayName: Use .NET
+        inputs:
+          packageType: 'sdk'
+          version: 8.0.300
```
Comment (CristianAmbrosini) [inline azure-pipelines.yml:255]: Educational: what is this about?

**Review by CristianAmbrosini — CHANGES_REQUESTED** [review]
Left few comments/questions

File: analyzers/tests/SonarAnalyzer.Test/TestCases/Utilities/MetricsAnalyzer/Razor.razor (hunk 5 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/Utilities/MetricsAnalyzer/Razor.razor:2]: They were false negatives that now are fixed. See in the `MetricsAnalyzerTest.cs` the new detected lines (1 and 2).

File: azure-pipelines.yml (hunk 8 above)
Comment (CristianAmbrosini) [reply azure-pipelines.yml:255]: You answered [here](https://github.com/SonarSource/sonar-dotnet/pull/9295#discussion_r1604937061). All good on my side

File: analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/MetricsAnalyzerTest.cs (hunk 4 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/MetricsAnalyzerTest.cs:72]: I preserved the order in which the lines were detected. `BeEquivalentTo` is not requiring that so I can change the order, indeed :)

**Review by CristianAmbrosini — APPROVED** [review]
LGTM!

**Pull request merged**